### PR TITLE
[SYCLomatic] Optimize cudaMemcpy migration

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

**Comment by jinz2014** [comment]
It may be helpful to add comments about the sections in the CUDA programming guide where memcpy is asynchronous. 

I had assumed that "cudaMemcpy()" is always synchronous.

Thanks

File: clang/lib/DPCT/Diagnostics.inc [hunk 1]
```diff
@@ -241,3 +241,5 @@ DEF_WARNING(TEX_FETCH, 1112, "If the filter mode is set to 'linear', the behavio
 DEF_COMMENT(TEX_FETCH, 1112, "If the filter mode is set to 'linear', the behavior of \"read\" may be different from \"tex1Dfetch\". You may need to adjust the code.")
 DEF_WARNING(ONE_DIMENSION_KERNEL_BARRIER, 1113, "Consider replacing sycl::nd_item::barrier(sycl::access::fence_space::local_space) with sycl::nd_item::barrier() if function \"%0\" is called in a multidimensional kernel.")
 DEF_COMMENT(ONE_DIMENSION_KERNEL_BARRIER, 1113, "Consider replacing sycl::nd_item::barrier(sycl::access::fence_space::local_space) with sycl::nd_item::barrier() if function \"{0}\" is called in a multidimensional kernel.")
+DEF_WARNING(WAIT_REMOVE, 1114, "There is no synchronize operation append after memcpy. This is based on the assumption that the original cudaMemcpy operations are asynchronous when data is transferred to the device and the memory involved is neither Unified Memory nor pinned memory. If this assumption doesn't hold, please append .wait() after memcpy for correct synchronization.")
```
Comment (zhiweij1) [inline clang/lib/DPCT/Diagnostics.inc]: In https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-sync , it only mentioned the terminologies "pageable memory" and "pinned memory" but no "Unified Memory".
Is it confusing that we mention "Unified Memory" here?

Comment (zhiweij1) [inline clang/lib/DPCT/Diagnostics.inc]: Although `cudaMemcpy` may perform as an asynchronous API, as mentioned in https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gc263dbe6574220cc776b45438fc351e8 , it "exhibits synchronous behavior for **most use cases**".
So does it a little strange we make the migration to be asynchronous? (Although --optimize-migration is on)

File: clang/test/dpct/memcpy_optimization.cu [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// RUN: dpct --format-range=none --optimize-migration -out-root %T/memcpy_optimization %s --cuda-include-path="%cuda-path/include" -- -x cuda --cuda-host-only
+// RUN: FileCheck --match-full-lines --input-file %T/memcpy_optimization/memcpy_optimization.dp.cpp %s
+#include <cuda_runtime.h>
+
+int main(){
+
+    float *a, *b;
+//CHECK:  /*
+//CHECK:  DPCT1114:{{[0-9]+}}: There is no synchronization operation append after memcpy. This is based on the assumption that the source host memory is pageable. If this assumption doesn't hold, please append .wait() after memcpy for correct synchronization behavior.
```
Comment (zhimingwang36) [inline clang/test/dpct/memcpy_optimization.cu]: cudaMemcpy is migrated to asynchronization memcpy, assuming in original code the source host memory is pageable memory. If this assumption doesn't hold, please append ‘.wait()’ after memcpy to  synchronize.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]

**Pull request merged**